### PR TITLE
Use Firedrake's static condensation preconditioner in the hybridized solver

### PR DESCRIPTION
**Pull request by thomasgibson**
This PR updates the linear solvers in Gusto. This includes a small QOL change that makes all time-stepping coefficients `ufl.Constant`s. This avoids repeated code-generation when all you want to do is change the time-step parameter.

More importantly, the hybridized compressible solver has a been refactored and organized. Rather than hand-coding the Slate expressions in the solver, we can do the static condensation directly via Firedrake's tested preconditioner `firedrake.SCPC`. This avoids a lot of potential mistakes that could be made when writing out the condensation operators by hand.

In fact, it turns out using `SCPC` has resolved some issues previously encountered by @tommbendall. Most notably, it appears that #169 has been fixed when using this branch.

Also, the code is much simpler to read I think.

### CODE REVIEW

File: gusto/state.py [hunk 1]
```diff
@@ -326,14 +326,13 @@ def setup_dump(self, t, tmax, pickup=False):
...
-            else:
-                if not running_tests:
-                    makedirs(self.dumpdir)
+            if self.mesh.comm.rank == 0:
+                if not running_tests and path.exists(self.dumpdir) and not pickup:
+                    raise IOError("results directory '%s' already exists"
+                                  % self.dumpdir)
+                else:
+                    if not running_tests:
+                        makedirs(self.dumpdir)
```
Comment (thomasgibson) [inline gusto/state.py]: This bit here fixes #190. I just left it in this PR since it was a very small change to the existing logic.

Comment (jshipton) [reply gusto/state.py]: Great - thanks! It's be good to have this in separately if the test fixes prove tricky...

File: gusto/linear_solvers.py [hunk 2]
```diff
@@ -306,37 +323,42 @@ def __init__(self, state, quadrature_degree=None, solver_parameters=None,
                 state.logger.warning("default quadrature degree most likely not sufficient for this degree element")
             self.quadrature_degree = (5, 5)
 
+        # Turn monitor on for the trace system when running in debug mode
+        if state.output.log_level == DEBUG:
```
Comment (jshipton) [inline gusto/linear_solvers.py]: I just merged a PR that has replaced all the instances of `if state.output.log_level == DEBUG` with `if logger.inEnabledFor(DEBUG)`. I think that in the future we will not have to pass state around so much...

Comment (thomasgibson) [reply gusto/linear_solvers.py]: Yep I noticed! I'm fixing it locally now as I'm addressing some Jenkins shenanigans

**Comment by thomasgibson** [comment]
Okay something is very strange. I'm passing the tests which "fail" on Jenkins using my laptop (macbook). In fact, @tommbendall reports problems on Ubuntu with this branch. Example `dry_bf_bubble.py` is one test where it fails on Jenkins and Ubuntu but works just fine on mac. I have no idea how to proceed from here.

**Comment by dham** [comment]
Try logging in to pex.ma.ic.ac.uk and running tests there. That's an ubuntu machine so you ought to at least be able to reproduce the issue.

**Comment by thomasgibson** [comment]
OK, green lights on Jenkins now!

**Pull request merged**